### PR TITLE
Implement updated model structure

### LINKED ISSUE #363: New design of model structure
I am thinking about redesigning the model structure of Chainer. It aims at following goals:

- Provide "the standard way" to define a structured model
- Provide a safe and flexible way to save and load models
- Make Optimizer more flexible

I wrote a document on a new design:
https://docs.google.com/document/d/1iO30OCnMFvn48Q6-2BOgKsJQXIKe0EQuCGwRdDJlgaE/edit#.
In this document I wrote several other issues that I want to resolve by the new design. I will try to implement it in a new branch as soon as possible. Please add comments on the document or this issue if you have any questions, suggestions, or discussion items.

### PR DESCRIPTION
**Pull request by beam2d**
Fixes #363. This PR is a replacement of #459 under the renewed design doc.

It adds following changes:
- Add Link, Chain, and ChainList classes for new style model construction
- Serialization protocol with HDF5 support (via h5py package)
- Add chainer.links namespace and some links under it
- Each parameterized function is replaced by corresponding Link implementation and moved to chainer.links (while also left at chainer.functions for backward compatibility)
- Optimizer interface is changed (backward compatibility is kept as much as possible)
- FunctionSet is reimplemented as a child class of Chain (backward compatibility is kept as much as possible)
- Function is now not allowed to have a state (including parameters)
- Split special function is removed
- Add Flag class for ternary volatility flag (ON, OFF, and AUTO)

I'm still working on updates of tutorials and examples. I will remove the WIP mark after these tasks are done.

### CODE REVIEW

File: docs/source/tutorial/basic.rst [hunk 1]
```diff
@@ -308,40 +399,55 @@ First, we scale pixels to [0, 1] values, and divide the dataset into 60,000 trai
...
-:func:`functions.softmax_cross_entropy` computes the loss function of softmax regression.
-:func:`functions.accuracy` computes the classification accuracy of this minibatch.
+.. doctest::
+
+   >>> model = L.Classifier(MLP())
+   >>> optimizer = optimizers.SGD()
+   >>> optimizer.setup(model)
 
-Finally, we can write a learning loop as following:
+Finally, we can write a leraning loop as following:
```
Comment (unnonouno) [inline docs/source/tutorial/basic.rst]: typo: leraning -> learning

**Comment by unnonouno** [comment]
OK, we'll first merge this PR, and then fix small problems.

**Comment by delta2323** [comment]
I agree with @unnonouno's idea that we first merge this PR and mend the errors.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**